plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,305 @@
+          }
+    flip runReader env <$> runGenT g
+
+-- * Utility functions
+
+-- | Don't allow types to escape from a generator.
+noEscape :: GenTm a -> GenTm a
+noEscape = local $ \env -> env { geEscaping = NoEscape }
+
+-- * Dealing with size
+
+-- | Map a function over the generator size
+onSize :: (Int -> Int) -> GenTm a -> GenTm a
+onSize f = local $ \ env -> env { geSize = f (geSize env) }
+
+-- | Default to the first generator if the size is zero (or negative),
+-- use the second generator otherwise.
+ifSizeZero :: GenTm a -> GenTm a -> GenTm a
+ifSizeZero ifZ nonZ = do
+  n <- asks geSize
+  if n <= 0 then ifZ else nonZ
+
+-- | Locally set the size in a generator
+withSize :: Int -> GenTm a -> GenTm a
+withSize = onSize . const
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments.
+sizeSplit_ :: Int -> Int -> GenTm a -> GenTm b -> GenTm (a, b)
+sizeSplit_ a b ga gb = sizeSplit a b ga (const gb)
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments and use the result of the first generator
+-- in the second.
+sizeSplit :: Int -> Int -> GenTm a -> (a -> GenTm b) -> GenTm (a, b)
+sizeSplit a b ga gb = do
+  n <- asks geSize
+  let na = (a * n) `div` (a + b)
+      nb = (b * n) `div` (a + b)
+  x <- withSize na ga
+  (,) x <$> withSize nb (gb x)
+
+-- * Names
+
+-- | Get all uniques we have generated and are used in the current context.
+getUniques :: GenTm (Set Unique)
+getUniques = do
+  GenEnv{geDatas = dts, geTypes = tys, geTerms = tms, geUnboundUsedTyNames = used} <- ask
+  return $ Set.mapMonotonic (_nameUnique . unTyName) (Map.keysSet dts <> Map.keysSet tys <> used) <>
+           Set.mapMonotonic _nameUnique (Map.keysSet tms) <>
+           Set.unions [ names d | d <- Map.elems dts ]
+  where
+    names (Datatype _ _ _ m cs) = Set.fromList $ _nameUnique m : [ _nameUnique c | VarDecl _ c _ <- cs ]
+
+{- Note [Warning about generating fresh names]: because `GenTm` is a *reader* monad
+   names are not immediately put into any state when generated. There is *no guarantee*
+   that in this situation:
+   ```
+   do nms <- genFreshNames ss
+      nms' <- genFreshNames ss
+   ```
+   the names in `nms` and `nms'` don't overlap.
+
+   Instead, what you are supposed to do is locally use the names in `nms` and `nms'` to
+   define generators that use them. This is done with functions like `bindTyName` and `bindTmName`:
+   ```
+   genLam ma mb = do
+      x <- genFreshName "x"
+      sizeSplit 1 7 (maybe (genType Star) return ma)
+                    --      v--- LOOK HERE!
+                    (\ a -> bindTmName x a . noEscape $ genTerm mb) $ \ a (b, body) ->
+                    --      ^--- LOOK HERE!
+                    TyFun () a b, LamAbs () x a body)
+   ```
+-}
+
+-- | Generate a fresh name. See Note [Warning about generating fresh names].
+genFreshName :: String -> GenTm Name
+genFreshName s = head <$> genFreshNames [s]
```

Comment:
Right, this might be a little bit slow. But I think we discussed the state solution elsewhere and it doesn't quite work out right? But `QuoteT` might still work for names though - even if we have `Reader`-style binders.

Reply:
I've played with it and realized that simply sticking `QuoteT` into the monad stack isn't gonna cut it, since the current fresh name generation logic is pretty fancy and we want to keep the fanciness, since it allows us to generate diverse and interesting terms. I don't think it makes any sense to declare this as a blocker, so I just created a ticket in Jira and gonna resolve this conversation.